### PR TITLE
D003 paging

### PR DESCRIPTION
**Pull request by DilyanPetkov**

### CODE REVIEW

File: client-service/src/main/java/com/registry/clientservice/controller/ClientController.java [hunk 1]
```diff
@@ -15,17 +15,18 @@
...
-    public Client getClientByClientNumber(@PathVariable("clientNumber") String clientNumber){
+    public ClientDTO getClientByClientNumber(@PathVariable("clientNumber") String clientNumber){
         return clientService.findOneByClientNumber(clientNumber);
     }
 
-    @GetMapping("/clients")
-    public List<Client> getAllClients(){
-        return clientService.getAllClients();
+    @GetMapping
+    public Page<Client> getAllClients(@RequestParam(value = "page", required = false) Integer page,
```
Comment (SavageStan) [inline client-service/src/main/java/com/registry/clientservice/controller/ClientController.java]: Is there any reason not to use ClientDTO here as well?

File: client-service/src/main/java/com/registry/clientservice/dto/ClientDTO.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.registry.clientservice.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class ClientDTO {
```
Comment (SavageStan) [inline client-service/src/main/java/com/registry/clientservice/dto/ClientDTO.java:10]: The id of the client should probably be included here as well?

File: client-service/src/main/java/com/registry/clientservice/service/ClientService.java [hunk 3]
```diff
@@ -1,28 +1,40 @@
...
 
-    public Client findOneByClientNumber(String clientNumber){
-        return clientRepository.findByClientNumber(clientNumber);
+    public ClientDTO findOneByClientNumber(String clientNumber) {
+        return mapToDTO(clientRepository.findByClientNumber(clientNumber));
+    }
+
+    public Page<Client> getAllClients(Integer page, Integer size) {
+        if (page == null || size == null) {
+            return clientRepository.findAll(PageRequest.of(0, (int) clientRepository.count(), Sort.by(Sort.Direction.ASC, "clientName")));
```
Comment (SavageStan) [inline client-service/src/main/java/com/registry/clientservice/service/ClientService.java]: What will happen here if the number of client entities in clientRepository is above int max value?

File: client-service/src/main/java/com/registry/clientservice/service/ClientService.java [hunk 4]
```diff
@@ -1,28 +1,40 @@
...
 
-    public List<Client> getAllClients(){
-        return clientRepository.findAll();
+    public ClientDTO createClient(Client client) {
+        return mapToDTO(clientRepository.save(client));
     }
 
-    public Client createClient(Client client){
-        return clientRepository.save(client);
+    public ClientDTO mapToDTO(Client client){
```
Comment (SavageStan) [inline client-service/src/main/java/com/registry/clientservice/service/ClientService.java]: Does this method need to be public?

File: cloud-gateway/src/main/java/com/registry/cloudgateway/controller/CloudGatewayController.java [hunk 5]
```diff
@@ -22,28 +25,35 @@ public ClientDTO getClientById(@PathVariable("id") String clientNumber) {
...
+        return cloudGatewayService.getAllInvoices(page, size);
     }
 
     @PostMapping("/invoices")
     public InvoiceDTO createInvoice(@RequestBody InvoiceDTO invoiceDTO) {
         return cloudGatewayService.createInvoice(invoiceDTO);
     }
 
+    @PostMapping("/invoices/search")
+    public List<InvoiceDTO> searchInvoice(@RequestBody CriteriaDTO criteriaDTO) {
```
Comment (SavageStan) [inline cloud-gateway/src/main/java/com/registry/cloudgateway/controller/CloudGatewayController.java]: I would call this DTO something less generic. For example InvoiceSearchRequestDTO.

File: cloud-gateway/src/main/java/com/registry/cloudgateway/dto/CriteriaDTO.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.math.BigDecimal;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CriteriaDTO {
+    BigDecimal quantity;
```
Comment (SavageStan) [inline cloud-gateway/src/main/java/com/registry/cloudgateway/dto/CriteriaDTO.java]: Does it make sense to use BigDecimal for quantity? Is the reason for this just to make computations easier (require no conversions)?

File: cloud-gateway/src/main/java/com/registry/cloudgateway/dto/ItemDTO.java [hunk 7]
```diff
@@ -13,9 +15,9 @@
     private String serialNumber;
     private String catalogueNumber;
     private String name;
-    private Double singlePrice;
-    private Double vat;
-    private Double quantity;
-    private Double totalPrice;
+    private BigDecimal singlePrice;
+    private BigDecimal vat;
+    private BigDecimal quantity;
```
Comment (SavageStan) [inline cloud-gateway/src/main/java/com/registry/cloudgateway/dto/ItemDTO.java]: Same question for this quantity field.

File: client-service/src/main/java/com/registry/clientservice/dto/ClientDTO.java (hunk 2 above)
Comment (DilyanPetkov) [reply client-service/src/main/java/com/registry/clientservice/dto/ClientDTO.java:10]: Nope, i don't use the id in any controller/service method.

File: cloud-gateway/src/main/java/com/registry/cloudgateway/dto/RestResponsePage.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class RestResponsePage<T> extends PageImpl<T> {
+
+    private static final long serialVersionUID = 3248189030448292002L;
```
Comment (SavageStan) [inline cloud-gateway/src/main/java/com/registry/cloudgateway/dto/RestResponsePage.java]: I don't think you need this serialVersionUID. Fields like this one are used when you are using Java serialization. This class does not implement Serializable, so I don't think we need it. Jackson uses a different kind of serialization (object to JSON).

File: cloud-gateway/src/main/java/com/registry/cloudgateway/dto/CriteriaDTO.java (hunk 6 above)
Comment (DilyanPetkov) [reply cloud-gateway/src/main/java/com/registry/cloudgateway/dto/CriteriaDTO.java]: Yes, i did it for computations mainly, quantity should not be like this, but changing it again would require changes again in the DB

File: cloud-gateway/src/main/java/com/registry/cloudgateway/dto/RestResponsePage.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public RestResponsePage(@JsonProperty("content") List<T> content, @JsonProperty("number") int number, @JsonProperty("size") int size,
```
Comment (SavageStan) [inline cloud-gateway/src/main/java/com/registry/cloudgateway/dto/RestResponsePage.java:16]: Considering that the variables from the parameter list of this method have the same names as the attributes of their corresponding @JsonProperty annotations, do we even need the annotations?

File: invoice-service/src/main/java/com/registry/invoiceservice/InvoiceServiceApplication.java [hunk 10]
```diff
@@ -3,10 +3,6 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.cloud.netflix.eureka.EnableEurekaClient;
-import org.springframework.context.annotation.Bean;
-import org.springframework.web.client.RestTemplate;
-
-import java.beans.BeanProperty;
 
 @SpringBootApplication
 @EnableEurekaClient
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/InvoiceServiceApplication.java]: Can we try if everything works without any Eureka related annotations and config (if any is left).

File: invoice-service/src/main/java/com/registry/invoiceservice/controller/InvoiceController.java [hunk 11]
```diff
@@ -15,18 +18,24 @@
...
+    @GetMapping
+    public Page<Invoice> findAllInvoices(@RequestParam(value = "page", required = false) Integer page,
+                                         @RequestParam(value = "size", required = false) Integer size) {
+        return invoiceService.getAllInvoices(page, size);
     }
 
-    @PostMapping("/invoices")
-    public Invoice createInvoice(@RequestBody Invoice invoice){
+    @PostMapping
+    public InvoiceDTO createInvoice(@RequestBody Invoice invoice) {
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/controller/InvoiceController.java]: Endpoints should not be using entities as request bodies.

File: invoice-service/src/main/java/com/registry/invoiceservice/dto/CriteriaDto.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.math.BigDecimal;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CriteriaDto {
+    BigDecimal quantityGreaterThan;
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/dto/CriteriaDto.java]: Same quantity related question (BigDecimal).

File: invoice-service/src/main/java/com/registry/invoiceservice/dto/ItemDTO.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+@NoArgsConstructor
+@Data
+public class ItemDTO {
+
+    private String serialNumber;
+    private String catalogueNumber;
+    private String name;
+    private BigDecimal singlePrice;
+    private BigDecimal vat;
+    private BigDecimal quantity;
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/dto/ItemDTO.java]: Same quantity related question.

File: invoice-service/src/main/java/com/registry/invoiceservice/entity/Invoice.java [hunk 14]
```diff
@@ -11,11 +11,12 @@
 @AllArgsConstructor
 @NoArgsConstructor
 @Data
+@Table(name = "invoice")
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/entity/Invoice.java]: I think you don't need the @Table annotation if name attribute has the same value as the name of your entity.

File: invoice-service/src/main/java/com/registry/invoiceservice/entity/Item.java [hunk 15]
```diff
@@ -17,14 +18,14 @@
...
     private String serialNumber;
     private String catalogueNumber;
     private String name;
-    private Double singlePrice;
-    private Double vat;
-    private Double quantity;
-    private Double totalPrice;
+    private BigDecimal singlePrice;
+    private BigDecimal vat;
+    private BigDecimal quantity;
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/entity/Item.java]: Same quantity related question.

File: invoice-service/src/main/java/com/registry/invoiceservice/service/InvoiceService.java [hunk 16]
```diff
@@ -15,21 +20,61 @@
...
+        itemDTO.setQuantity(item.getQuantity());
+        itemDTO.setCatalogueNumber(item.getCatalogueNumber());
+        itemDTO.setTotalPrice(item.getTotalPrice());
+        itemDTO.setSerialNumber(item.getSerialNumber());
+        itemDTO.setSinglePrice(item.getSinglePrice());
+        itemDTO.setVat(item.getVat());
+        return itemDTO;
+    }
+
+    private InvoiceDTO mapToInvoiceDTO(Invoice invoice){
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/service/InvoiceService.java:92]: If we have time (in terms of training plan) we should look into mapping libraries like MapStruct.

File: invoice-service/src/main/java/com/registry/invoiceservice/controller/InvoiceController.java [hunk 17]
```diff
@@ -15,18 +18,24 @@
...
-    public Invoice getById(@PathVariable("id") Integer id){
+    public InvoiceDTO getById(@PathVariable("id") Long id) {
         return invoiceService.getInvoiceById(id);
     }
 
-    @GetMapping("/invoices")
-    public List<Invoice> findAllInvoices(){
-        return invoiceService.getAllInvoices();
+    @GetMapping
+    public Page<Invoice> findAllInvoices(@RequestParam(value = "page", required = false) Integer page,
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/controller/InvoiceController.java]: Entity -> DTO

File: invoice-service/src/main/java/com/registry/invoiceservice/dto/CriteriaDto.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.math.BigDecimal;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CriteriaDto {
```
Comment (SavageStan) [inline invoice-service/src/main/java/com/registry/invoiceservice/dto/CriteriaDto.java]: Name this CriteriaDTO. I have a suggestion for an even better name InvoiceSearchRequestDTO.

**Review by SavageStan — APPROVED** [review]

**Pull request merged**